### PR TITLE
nodejs: fix cross compilation for armv7l

### PR DESCRIPTION
**Pull request by Cynerd**
###### Description of changes
The cross build for armv7l requires build to be on i686 (32bit in general). This ensures that and thus fixes build for armv7l.

I am not happy with this solution but I am unable to figure out how to fix cross compilation for amv7l in better way than this.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by alyssais — CHANGES_REQUESTED** [review]
If we truly need to force a 32-bit build system, I think a cleaner way to do it would be to do it in the `callPackage` invocation.  Then we could just do something like

```
nodejs = (if stdenv.buildPlatform.isx86_64 && stdenv.is32bit then pkgsi686Linux else pkgs).callPackage …
```

which is nice because then there's one less package set involved in the actual expression, and it prevents adding 64-bit dependencies as well by mistake.

Either way, this needs a comment explaining what's going on, with a link to documentation or reference to the source code so people can understand what's going on, and where to check to see if it can be removed.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (alyssais) [inline pkgs/development/web/nodejs/nodejs.nix:44]: (Small note, that will be irrelevant if we use the callPackage version instead.) This is incorrect.  buildPackages is equivalent to nativeBuildInputs.  For depsBuildBuild, packages are sourced from pkgsBuildBuild.

**Comment by Cynerd** [comment]
> If we truly need to force a 32-bit build system, I think a cleaner way to do it would be to do it in the callPackage invocation. Then we could just do something like

I tweaked it a bit. It is probably reasonable requirement.

> Either way, this needs a comment explaining what's going on, with a link to documentation or reference to the source code so people can understand what's going on, and where to check to see if it can be removed.

I added comment but I am unable to find that this is actually required in the documentation. At the same time build really fails with missing 32bit stubs and thus it for sure tries to use 32bit build somewhere inside. The upstream builds it on Ubuntu and thus they might expect that 32bit libraries are just available while that is not the case with nixpkgs.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (Cynerd) [reply pkgs/development/web/nodejs/nodejs.nix]: Call package version used instead.

**Review by alyssais — APPROVED** [review]
LGTM but I'd really like there to be an upstream issue about this.  They shouldn't be passing -m32 to the native compiler just because they're cross compiling for 32 bit.

**Comment by lilyinstarlight** [comment]
I only just now saw this, but I do wonder if it may also be fixed by #238377?

**Comment by Cynerd** [comment]
@lilyinstarlight  I just tested it, and it doesn't help my issue. The issue here is with cross-compilation from amd64 to 32-bit systems. Node requires bitness to match it seems, and for that reason, we need to build it not in an amd64 environment but instead in an x86 one.

I suspect that, especially on new Macs that do not have 32-bit arm support (if I remember correctly) it won't even be possible to cross-compile Node to armv7l.

**Comment by Cynerd** [comment]
> LGTM but I'd really like there to be an upstream issue about this. They shouldn't be passing -m32 to the native compiler just because they're cross compiling for 32 bit.

I create an issue upstream but I am not expecting much https://github.com/nodejs/node/issues/48612. This is a common issue I saw with other bytecode compilers (for example luajit).

**Comment by tie** [comment]
FWIW I think this is the relevant code that sets -m32 when cross-compiling in your case.
https://github.com/nodejs/node/blob/3205b1936a5c3c21b0c8c568a57b0d2bada366b1/tools/v8_gypfiles/toolchain.gypi#L583-L601

host_cxx_is_biarch is defined at https://github.com/nodejs/node/blob/3205b1936a5c3c21b0c8c568a57b0d2bada366b1/tools/v8_gypfiles/toolchain.gypi#L107-L118

That is, it assumes that x86(-64) (and some other systems) have bi-arch capable toolchains (probably for a reason, see below, but maybe not, this is technical decision is really weird).

With the following patch, we successfully get a meaningful error (instead of vague compilation failure due to the missing headers) that leads us to https://github.com/nodejs/node/blob/3205b1936a5c3c21b0c8c568a57b0d2bada366b1/deps/v8/include/v8config.h#L863-L892 (or `deps/v8/src/base/build_config.h` for v18)

<details>

```patch
diff --git a/tools/v8_gypfiles/toolchain.gypi b/tools/v8_gypfiles/toolchain.gypi
index 26c6866fa9..658ebd6e86 100644
--- a/tools/v8_gypfiles/toolchain.gypi
+++ b/tools/v8_gypfiles/toolchain.gypi
@@ -109,7 +109,7 @@
       host_arch=="s390x" or \
       clang==1', {
       'variables': {
-        'host_cxx_is_biarch%': 1,
+        'host_cxx_is_biarch%': 0,
        },
      }, {
       'variables': {
```
</details>

> *Target architecture arm is only supported on arm and ia32 host*

So this seems like an issue with v8 rather than Node.js. At this point I don’t really want to dive deeper into this rabbit hole 🥲

File: pkgs/top-level/all-packages.nix
Comment (tie) [inline pkgs/top-level/all-packages.nix:9409]:
```suggestion
  # only architecture that generally supports execution of 32bit is x86_64 and thus that
```

IIRC AArch64 can execute AArch32 code if enabled in the kernel configuration ([Kernel support for 32-bit EL0 found in arch/arm64/Kconfig](https://cateee.net/lkddb/web-lkddb/COMPAT.html))? That seems to be CPU-dependent for ARM, though I think 32-bit can also be disabled on x86-64.

**Review by tie — APPROVED** [review]

**Comment by lilyinstarlight** [comment]
@Cynerd Could you rebase and resolve merge conflicts on this?

If this is what we have to do, I'm mostly alright with this, but it is rather unfortunate

File: pkgs/top-level/all-packages.nix
Comment (Cynerd) [reply pkgs/top-level/all-packages.nix]: Yes, that is CPU-dependent on aarch64 because it relies on the CPU to actually interpret those arm instructions, as far as I know. The only way I know is to use binfmt and register an emulator. On the second note about disable of x86: you can even theoretically disable the https://www.kernelconfig.io/CONFIG_COMPAT_BINFMT_ELF?kernelversion=6.3.12 it just depends on where you draw the line between "this is expected platform support and this is not". Linux kernel provides way too many compilations knobs to cover all corners.

But I agree with the comment change (and it is done by now). This answer is only just a polemic.

**Comment by Cynerd** [comment]
@lilyinstarlight rebase is done.

File: pkgs/top-level/all-packages.nix
Comment (tie) [inline pkgs/top-level/all-packages.nix:10014]: This is a public attribute that people may use even though it’s an internal implementation detail. I think it makes sense to move it to the `v{18,20,22,…}.nix` files (or at least avoid exposing it).

Comment (Cynerd) [reply pkgs/top-level/all-packages.nix:10014]: That was a previous implementation that I changed on bases of these arguments https://github.com/NixOS/nixpkgs/pull/220204#pullrequestreview-1347419337.

**Comment by tie** [comment]
Hm, I’ve just noticed how buildroot handles that, would it make sense to use `stdenv.hostPlatform.emulator` here? Is that even possible?

https://github.com/buildroot/buildroot/blob/ecf060a2ff93626fb5fbe1c52b8d42b6ee0fbb5b/package/nodejs/nodejs-src/nodejs-src.mk#L175-L204

File: pkgs/top-level/all-packages.nix
Comment (tie) [reply pkgs/top-level/all-packages.nix:10014]: >the standard callPackage is publicly available
Yes, because it is a generic public interface while `callPackageNodejs` is an alias used only to build Node.js package. Is there some other use case I’m missing here?

**Comment by tie** [comment]
>Although it might be required in the end considering that node closed the issue about support for cross compilation on different bitness and the need to also build it on not only x86 (amd64) architecture.

The change in this PR kinda overrides localSystem arguments of the Nixpkgs when building Node.js (see pkgsi686Linux definition in pkgs/top-level/stage.nix) and I’m not really comfortable with that, also https://github.com/NixOS/nixpkgs/pull/220204#discussion_r1256713670

>I am not sure how to transfer that directly to the nixpkgs because I would probably need to switch build from cross-compilation, but that would use all tools from the host platform.

Hm, at least getting the QEMU wrapper patch seems doable unless I’m missing something here.

- `v8-qemu-wrapper` translates directly to `stdenv.hostPlatform.emulator buildPackages`,
- `{HOST_,}NODEJS_SRC_WRAPPER_FIXUP` goes to the patchPhase (along with the accompanying `0001-add-qemu-wrapper-support.patch` in that directory).
- I don’t see any other unusual configure flags in buildroot package aside from them always using `--cross-compiling` and related flags (on that note, see also #317109 to avoid unnecessary rebases if someone with commit permissions finally sees it).

File: pkgs/top-level/all-packages.nix
Comment (Cynerd) [reply pkgs/top-level/all-packages.nix:10014]: Like packaging a custom version in overlays?

Comment (tie) [reply pkgs/top-level/all-packages.nix:10014]: Isn’t `override{,Attrs}` enough in this case?

Comment (Cynerd) [reply pkgs/top-level/all-packages.nix:10014]: You are right.

I will move it to the vxx.nix files.

**Comment by Cynerd** [comment]
> Hm, at least getting the QEMU wrapper patch seems doable unless I’m missing something here.

Yes, that is the easy part. The nasty part is in providing the host toolchain as a build one without actually using `callPackage` from the host platform (that would pull all dependencies and thus be effectively built in the qemu, which is really slow).

**Comment by tie** [comment]
I don’t see how that would be slow given that only a handful of build steps actually need an emulator.

**Comment by Cynerd** [comment]
> I don’t see how that would be slow given that only a handful of build steps actually need an emulator.

You misunderstood me. I was attempting to do that, and I was failing because the build system was mixing the host and build architectures.. at least, that is what I thought was happening. In the end, it turned out that I had an issue with the modification script, and it all went crashing down way later with some dependencies and errors about invalid architecture. My message was about, at that moment, seeing the only option to build it all in the host platform and thus emulate everything (that would be slow). In the end, I was able to solve it. I will push here the reworked version that uses an emulator to compile it. It started working once I fixed some stupid errors (missing commas).

**Comment by tie** [comment]
Yes, should be targeting staging given the number of rebuilds.

See https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging

**Comment by Cynerd** [comment]
Rebase and retargeting is done (I am not sure if GitHub sends an email about it).

File: pkgs/development/web/nodejs/add-qemu-wrapper-support.patch
Comment (tie) [inline pkgs/development/web/nodejs/add-qemu-wrapper-support.patch:1]: Are there any changes from the buildroot patch? Can we use `fetchpatch2` to fetch the patch?

File: pkgs/development/web/nodejs/nodejs.nix
Comment (tie) [inline pkgs/development/web/nodejs/nodejs.nix:131]: I’m not really a fan of conditional patches. Can we apply the patch unconditionally (and replace `MAYBE_WRAPPER` with an empty string if we are not using an emulator)?

File: pkgs/development/web/nodejs/nodejs.nix
Comment (tie) [inline pkgs/development/web/nodejs/nodejs.nix:158]: Hm, I think a symlink should be enough in this case. IIRC emulator contains arguments in addition to path iff `canExecute`.
https://github.com/NixOS/nixpkgs/blob/5d49679d00dedf21e4db29218c6ac27edad7855c/lib/systems/default.nix#L285-L295

File: pkgs/development/web/nodejs/nodejs.nix
Comment (tie) [inline pkgs/development/web/nodejs/nodejs.nix:65]: Are we using `$NIX_CC` instead of `$NIX_CC_FOR_BUILD` in this case? If we don’t need C compiler for build platform, perhaps it makes sense to omit `depsBuildBuild` entirely if `isCrossBitness`?

File: pkgs/development/web/nodejs/nodejs.nix
Comment (tie) [reply pkgs/development/web/nodejs/nodejs.nix:158]: Ah, OK.
```
nix-repl> (lib.systems.elaborate "x86_64-linux").canExecute (lib.systems.elaborate "i686-linux")   
true
```

**Comment by aduh95** [comment]
This needs a rebase.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (Cynerd) [reply pkgs/development/web/nodejs/nodejs.nix:158]: Yeah for i686 it needs to be a script.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (Cynerd) [reply pkgs/development/web/nodejs/nodejs.nix:131]: Ok.. in my eyes, it makes it a bit more complex, but ok. There is also an option to just always use a wrapper because, for the native build, it will work as well.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (Cynerd) [reply pkgs/development/web/nodejs/nodejs.nix:65]: Yes it should be possible to just drop `depsBuildBuild` in case of `isCrossBitness`.

Would it even make sense to always use this magic for cross-compilation and not just when bitness differed? The issue with `isCrossBitness` is that as far as I can see, this is only extraction for the Nix-supported platforms, but as you linked, v8 actually supports only limited build platforms not actually matching bitness.

File: pkgs/development/web/nodejs/add-qemu-wrapper-support.patch
Comment (Cynerd) [reply pkgs/development/web/nodejs/add-qemu-wrapper-support.patch:1]: No changes in the end. I will switch it to fetchpatch.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (tie) [reply pkgs/development/web/nodejs/nodejs.nix:158]: FWIW I’ve opened PR #324071 to make an emulator a simple executable path in native case.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (tie) [reply pkgs/development/web/nodejs/nodejs.nix:65]: >Would it even make sense to always use this magic for cross-compilation and not just when bitness differed?

Yes, sounds reasonable to me. Always using an emulator instead of cross-compiling should make Node.js build faster, see also #321847.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (Cynerd) [reply pkgs/development/web/nodejs/nodejs.nix:158]: With that it should be possible to use only just symlink 👍

File: pkgs/development/web/nodejs/nodejs.nix
Comment (Cynerd) [reply pkgs/development/web/nodejs/nodejs.nix]: It is now converted to build always with the patch and with wrapper.

I tested it on amd64 and cross build from amd64 to armv7.

File: pkgs/development/web/nodejs/nodejs.nix
Comment (Cynerd) [reply pkgs/development/web/nodejs/nodejs.nix]: Can we proceed with this PR and update it just once #324071 is merged. It should be later only a minor change in `postPath`. What do you think?

File: pkgs/development/web/nodejs/nodejs.nix
Comment (tie) [inline pkgs/development/web/nodejs/nodejs.nix:133]: GitHub repository is a mirror, the official Git repository is at https://gitlab.com/buildroot.org/buildroot.
```suggestion
        url = "https://gitlab.com/buildroot.org/buildroot/-/raw/ecf060a2ff93626fb5fbe1c52b8d42b6ee0fbb5b/package/nodejs/nodejs-src/0001-add-qemu-wrapper-support.patch";
```

File: pkgs/development/web/nodejs/nodejs.nix
Comment (tie) [inline pkgs/development/web/nodejs/nodejs.nix:198]: Hm, I’m not sure why this line is conditional. I thought Nixpkgs bintools wrapper handled response files, but maybe that’s not the case here for some reason? I’ll have to double-check that.

**Comment by tie** [comment]
@Cynerd, I’ve made a few minor adjustments based on this PR and re-implemented buildroot patch in 42fa83e8d79d78df8d603048b604c2a33cca479d ([0001-nodejs-fix-cross-compilation-for-armv7l.patch](https://github.com/user-attachments/files/16240864/0001-nodejs-fix-cross-compilation-for-armv7l.patch)) so that the emulator trick is upstreamable. What do you think?

**Comment by Cynerd** [comment]
> @Cynerd, I’ve made a few minor adjustments based on this PR and re-implemented buildroot patch in [42fa83e](https://github.com/NixOS/nixpkgs/commit/42fa83e8d79d78df8d603048b604c2a33cca479d) ([0001-nodejs-fix-cross-compilation-for-armv7l.patch](https://github.com/user-attachments/files/16240864/0001-nodejs-fix-cross-compilation-for-armv7l.patch)) so that the emulator trick is upstreamable. What do you think?

I am ok with that. I tried to compile it and ti seems to work so I would go with your solution.

**Comment by tie** [comment]
Note for reviewers: #327653 contains commits from this PR (with slight modifications, see above) and goes further on top of these changes to use ninja for build.

@Cynerd, would it be OK to close this PR (since you’ve [approved ninja PR](https://github.com/NixOS/nixpkgs/pull/327653#pullrequestreview-2185276941))?

**Comment by Cynerd** [comment]
@tie Yes. This can be closed.

**Pull request closed without merging**
